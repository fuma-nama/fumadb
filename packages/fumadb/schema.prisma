model users {
  id Int @id @default(autoincrement())
  name String
  email String
  image String? @default("my-avatar")
}

model accounts {
  id String @id
}

datasource db {
  provider = "sqlite"
  url      = env("DATABASE_URL")
}
generator client {
    provider = "prisma-client-js"
}