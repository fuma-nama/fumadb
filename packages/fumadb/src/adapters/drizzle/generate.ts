import { importGenerator } from "../../utils/import-generator";
import { ident, parseVarchar } from "../../utils/parse";
import { AnyColumn, AnySchema, AnyTable, IdColumn } from "../../schema/create";
import { SQLProvider } from "../../shared/providers";
import { schemaToDBType } from "../../schema/serialize";

export function generateSchema(
  schema: AnySchema,
  provider: Exclude<SQLProvider, "cockroachdb" | "mssql">
): string {
  const imports = importGenerator();
  const importSource = {
    mysql: "drizzle-orm/mysql-core",
    postgresql: "drizzle-orm/pg-core",
    sqlite: "drizzle-orm/sqlite-core",
  }[provider];

  const tableFn = {
    mysql: "mysqlTable",
    postgresql: "pgTable",
    sqlite: "sqliteTable",
  }[provider];

  const generatedCustomTypes = new Set<string>();
  function generateCustomType(
    name: string,
    options: {
      dataType: string;
      driverDataType: string;
      databaseDataType: string;

      fromDriverCode: string;
      toDriverCode: string;
    }
  ) {
    if (generatedCustomTypes.has(name)) return;

    imports.addImport("customType", importSource);
    generatedCustomTypes.add(name);
    return `const ${name} = customType<
  {
    data: ${options.dataType};
    driverData: ${options.driverDataType};
  }
>({
  dataType() {
    return "${options.databaseDataType}";
  },
  fromDriver(value) {
    ${options.fromDriverCode}
  },
  toDriver(value) {
    ${options.toDriverCode}
  }
});`;
  }

  function generateBinary() {
    const name = "customBinary";
    // most Node.js based drivers return Buffer for binary data, make sure to convert them
    const code = generateCustomType(name, {
      dataType: "Uint8Array",
      driverDataType: "Buffer",
      databaseDataType: schemaToDBType({ type: "binary" }, provider),
      fromDriverCode:
        "return new Uint8Array(value.buffer, value.byteOffset, value.byteLength)",
      toDriverCode: `return value instanceof Buffer? value : Buffer.from(value)`,
    });

    if (code) lines.push(code);
    return name;
  }

  function getColumnTypeFunction(column: AnyColumn): {
    name: string;
    isCustomType?: boolean;
    params?: string[];
  } {
    if (provider === "sqlite") {
      switch (column.type) {
        case "bigint":
          return {
            name: "blob",
            params: [`{ mode: "bigint" }`],
          };
        case "bool":
          return {
            name: "integer",
            params: [`{ mode: "boolean" }`],
          };
        case "json":
          return { name: "blob", params: [`{ mode: "json" }`] };
        // for sqlite, generate dates as a timestamp
        case "timestamp":
        case "date":
          return { name: "integer", params: [`{ mode: "timestamp" }`] };
        case "decimal":
          return { name: "real" };
      }
    }

    switch (column.type) {
      case "string":
        return { name: "text" };
      case "binary":
        return {
          name: generateBinary(),
          isCustomType: true,
        };
      case "bool":
        return { name: "boolean" };
      default:
        if (column.type.startsWith("varchar")) {
          return {
            name: provider === "sqlite" ? "text" : "varchar",
            params: [`{ length: ${parseVarchar(column.type)} }`],
          };
        }

        return { name: column.type };
    }
  }

  function generateTable(table: AnyTable) {
    const cols: string[] = [];

    for (const column of Object.values(table.columns)) {
      const col: string[] = [];
      const typeFn = getColumnTypeFunction(column);
      // Handle column type
      const params: string[] = [
        `"${column.names.sql}"`,
        ...(typeFn.params ?? []),
      ];

      if (!typeFn.isCustomType) imports.addImport(typeFn.name, importSource);
      col.push(`${typeFn.name}(${params.join(", ")})`);

      if (column instanceof IdColumn) {
        col.push("primaryKey()");

        if (column.default === "auto") {
          imports.addImport("createId", "fumadb/cuid");
          col.push("$defaultFn(() => createId())");
        }
      }

      if (column.unique) {
        col.push("unique()");
      }

      if (!column.nullable) {
        col.push("notNull()");
      }

      // Handle default values
      if (column.default === "now") {
        col.push("defaultNow()");
      } else if (typeof column.default === "object") {
        if ("sql" in column.default) {
          imports.addImport("sql", "drizzle-orm");
          col.push(`default(sql\`${column.default.sql}\`)`);
        } else {
          const value = JSON.stringify(column.default.value);
          col.push(`default(${value})`);
        }
      }

      cols.push(`  ${column.names.drizzle}: ${col.join(".")}`);
    }

    const args: string[] = [`"${table.names.sql}"`];
    args.push(`{\n${cols.join(",\n")}\n}`);

    const keys: string[] = [];
    for (const config of table.foreignKeys) {
      const referencedTable = schema.tables[config.referencedTable];

      const columns = config.columns.map(
        (col) => `table.${table.columns[col].names.drizzle}`
      );
      const foreignColumns = config.referencedColumns.map(
        (col) =>
          `${referencedTable.names.drizzle}.${referencedTable.columns[col].names.drizzle}`
      );

      imports.addImport("foreignKey", importSource);
      let code = `foreignKey({
  columns: [${columns.join(", ")}],
  foreignColumns: [${foreignColumns.join(", ")}],
  name: "${config.name}"
})`;
      if (config?.onUpdate)
        code += `.onUpdate("${config.onUpdate.toLowerCase()}")`;

      if (config?.onDelete)
        code += `.onDelete("${config.onDelete.toLowerCase()}")`;

      keys.push(code);
    }

    if (keys.length > 0)
      args.push(`(table) => [\n${ident(keys.join(",\n"))}\n]`);

    return `export const ${table.names.drizzle} = ${tableFn}(${args.join(", ")})`;
  }

  function generateRelation(table: AnyTable) {
    const cols: string[] = [];

    for (const relation of Object.values(table.relations)) {
      const options: string[] = [`relationName: "${relation.id}"`];

      // only `many` doesn't require fields, references
      if (!relation.implied || relation.type === "one") {
        const fields: string[] = [];
        const references: string[] = [];

        for (const [left, right] of relation.on) {
          fields.push(
            `${table.names.drizzle}.${table.columns[left].names.drizzle}`
          );
          references.push(
            `${relation.table.names.drizzle}.${relation.table.columns[right].names.drizzle}`
          );
        }

        options.push(
          `fields: [${fields.join(", ")}]`,
          `references: [${references.join(", ")}]`
        );
      }

      const args: string[] = [];
      args.push(relation.table.names.drizzle);
      if (options.length > 0) args.push(`{\n${ident(options.join(",\n"))}\n}`);

      cols.push(
        ident(`${relation.name}: ${relation.type}(${args.join(", ")})`)
      );
    }

    if (cols.length === 0) return;
    imports.addImport("relations", "drizzle-orm");
    return `export const ${table.names.drizzle}Relations = relations(${
      table.names.drizzle
    }, ({ one, many }) => ({
${cols.join(",\n")}
}));`;
  }

  imports.addImport(tableFn, importSource);
  const lines: string[] = [];
  for (const table of Object.values(schema.tables)) {
    lines.push(generateTable(table));
    const relation = generateRelation(table);
    if (relation) lines.push(relation);
  }

  lines.unshift(imports.format());
  return lines.join("\n\n");
}
