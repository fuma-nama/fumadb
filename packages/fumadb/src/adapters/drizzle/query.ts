import * as Drizzle from "drizzle-orm";
import { type SimplifyFindOptions, toORM } from "../../query/orm";
import type { AbstractQuery, FindManyOptions } from "../../query";
import { type AnyColumn, type AnySchema, type AnyTable, Column } from "../../schema";
import type { SQLProvider } from "../../shared/providers";
import { type Condition, ConditionType } from "../../query/condition-builder";
import type * as MySQL from "drizzle-orm/mysql-core";
import type * as PostgreSQL from "drizzle-orm/pg-core";

type TableType = MySQL.MySqlTableWithColumns<MySQL.TableConfig>;
type ColumnType = MySQL.AnyMySqlColumn;
type DBType = MySQL.MySqlDatabase<
  MySQL.MySqlQueryResultHKT,
  MySQL.PreparedQueryHKTBase,
  Record<string, unknown>,
  Drizzle.TablesRelationalConfig
>;

type P_TableType = PostgreSQL.PgTableWithColumns<PostgreSQL.TableConfig>;
type P_ColumnType = PostgreSQL.AnyPgColumn;
type P_DBType = PostgreSQL.PgDatabase<
  PostgreSQL.PgQueryResultHKT,
  Record<string, unknown>,
  Drizzle.TablesRelationalConfig
>;

function buildWhere(
  toDrizzle: (col: AnyColumn) => ColumnType,
  condition: Condition
): Drizzle.SQL | undefined {
  if (condition.type === ConditionType.Compare) {
    const left = toDrizzle(condition.a);
    const op = condition.operator;
    let right = condition.b;
    if (right instanceof Column) right = toDrizzle(right);
    let inverse = false;

    switch (op) {
      case "=":
        return Drizzle.eq(left, right);
      case "!=":
        return Drizzle.ne(left, right);
      case ">":
        return Drizzle.gt(left, right);
      case ">=":
        return Drizzle.gte(left, right);
      case "<":
        return Drizzle.lt(left, right);
      case "<=":
        return Drizzle.lte(left, right);
      case "in": {
        // @ts-expect-error -- skip type check
        return Drizzle.inArray(left, right);
      }
      case "not in":
        // @ts-expect-error -- skip type check
        return Drizzle.notInArray(left, right);
      case "is":
        return right === null ? Drizzle.isNull(left) : Drizzle.eq(left, right);
      case "is not":
        return right === null
          ? Drizzle.isNotNull(left)
          : Drizzle.ne(left, right);
      case "not contains":
        inverse = true;
      case "contains":
        right =
          typeof right === "string"
            ? `%${right}%`
            : Drizzle.sql`concat('%', ${right}, '%')`;

        return inverse
          ? // @ts-expect-error -- skip type check
            Drizzle.notLike(left, right)
          : // @ts-expect-error -- skip type check
            Drizzle.like(left, right);
      case "not ends with":
        inverse = true;
      case "ends with":
        right =
          typeof right === "string"
            ? `%${right}`
            : Drizzle.sql`concat('%', ${right})`;

        return inverse
          ? // @ts-expect-error -- skip type check
            Drizzle.notLike(left, right)
          : // @ts-expect-error -- skip type check
            Drizzle.like(left, right);
      case "not starts with":
        inverse = true;
      case "starts with":
        right =
          typeof right === "string"
            ? `${right}%`
            : Drizzle.sql`concat(${right}, '%')`;

        return inverse
          ? // @ts-expect-error -- skip type check
            Drizzle.notLike(left, right)
          : // @ts-expect-error -- skip type check
            Drizzle.like(left, right);

      default:
        throw new Error(`Unsupported operator: ${op}`);
    }
  }

  if (condition.type === ConditionType.And)
    return Drizzle.and(
      ...condition.items.map((item) => buildWhere(toDrizzle, item))
    );

  if (condition.type === ConditionType.Not) {
    const result = buildWhere(toDrizzle, condition.item);
    if (!result) return;

    return Drizzle.not(result);
  }

  return Drizzle.or(
    ...condition.items.map((item) => buildWhere(toDrizzle, item))
  );
}

function mapValues(
  values: Record<string, unknown>,
  table: AnyTable
): Record<string, unknown> {
  const out: Record<string, unknown> = {};

  for (const column of Object.values(table.columns)) {
    out[column.names.drizzle] = values[column.ormName];
  }

  return out;
}

function mapQueryResult(table: AnyTable, result: Record<string, unknown>) {
  const out: Record<string, unknown> = {};

  for (const k in result) {
    const value = result[k];

    if (k in table.relations) {
      const relation = table.relations[k];

      if (relation.type === "many") {
        out[k] = (value as Record<string, unknown>[]).map((v) =>
          mapQueryResult(relation.table, v)
        );
        continue;
      }

      out[k] = value ? mapQueryResult(relation.table, value as any) : null;
      continue;
    }

    const col = table.getColumnByName(k, "drizzle");
    if (!col) continue;
    out[col.ormName] = value;
  }

  return out;
}

// TODO: Support binary data in relation queries, because Drizzle doesn't support it: https://github.com/drizzle-team/drizzle-orm/issues/3497
/**
 * Require drizzle query mode, make sure to configure it first. (including the `schema` option)
 */
export function fromDrizzle(
  schema: AnySchema,
  _db: unknown,
  provider: SQLProvider
): AbstractQuery<AnySchema> {
  const db = _db as DBType;
  const drizzleTables = db._.fullSchema as Record<string, TableType>;
  if (!drizzleTables || Object.keys(drizzleTables).length === 0)
    throw new Error(
      "[fumadb] Drizzle adapter requires query mode, make sure to configure it following their guide: https://orm.drizzle.team/docs/rqb."
    );

  function toDrizzle(v: AnyTable): TableType {
    const out = drizzleTables[v.names.drizzle];
    if (out) return out;

    throw new Error(
      `[FumaDB Drizzle] Unknown table name ${v.names.drizzle}, is it included in your Drizzle schema?`
    );
  }

  function toDrizzleColumn(v: AnyColumn): ColumnType {
    const table = toDrizzle(v._table!);
    const out = table[v.names.drizzle];
    if (out) return out;

    throw new Error(
      `[FumaDB Drizzle] Unknown column name ${v.names.drizzle} in ${v._table?.names.drizzle}.`
    );
  }

  // Drizzle Queries doesn't support renaming fields with `mapWith` because https://github.com/drizzle-team/drizzle-orm/issues/1157
  // we need to map the result on JS instead of relying on Drizzle
  function buildQueryConfig(
    table: AnyTable,
    options: SimplifyFindOptions<FindManyOptions>
  ) {
    const columns: Record<string, boolean> = {};
    const select = options.select;

    if (select === true) {
      for (const col of Object.values(table.columns)) {
        columns[col.names.drizzle] = true;
      }
    } else {
      for (const k of select) {
        columns[table.columns[k].names.drizzle] = true;
      }
    }

    const out: Drizzle.DBQueryConfig<"many" | "one", boolean> = {
      columns,
      limit: options.limit,
      offset: options.offset,
      where: options.where
        ? buildWhere(toDrizzleColumn, options.where)
        : undefined,
      orderBy: options.orderBy?.map(([item, mode]) =>
        mode === "asc"
          ? Drizzle.asc(toDrizzleColumn(item))
          : Drizzle.desc(toDrizzleColumn(item))
      ),
    };

    if (options.join) {
      out.with = {};

      for (const join of options.join) {
        if (join.options === false) continue;

        out.with[join.relation.name] = buildQueryConfig(
          join.relation.table,
          join.options
        );
      }
    }

    return out;
  }

  return toORM({
    tables: schema.tables,
    async count(table, v) {
      return await db.$count(
        toDrizzle(table),
        v.where ? buildWhere(toDrizzleColumn, v.where) : undefined
      );
    },
    async findFirst(table, v) {
      const results = await this.findMany(table, {
        ...v,
        limit: 1,
      });

      return results[0] ?? null;
    },

    async upsert(table, v) {
      const idField = table.getIdColumn().names.drizzle;
      const drizzleTable = toDrizzle(table);
      let query = db
        .select({ id: drizzleTable[idField] })
        .from(drizzleTable)
        .limit(1);

      if (v.where) {
        query = query.where(buildWhere(toDrizzleColumn, v.where)) as any;
      }

      const targetIds = await query.execute();

      if (targetIds.length > 0) {
        await db
          .update(drizzleTable)
          .set(mapValues(v.update, table))
          .where(Drizzle.eq(drizzleTable[idField], targetIds[0].id));
      } else {
        await this.createMany(table, [v.create]);
      }
    },
    async findMany(table, v) {
      return (
        await db.query[table.names.drizzle].findMany(buildQueryConfig(table, v))
      ).map((v) => mapQueryResult(table, v));
    },

    async updateMany(table, v) {
      const drizzleTable = toDrizzle(table);

      let query = db.update(drizzleTable).set(mapValues(v.set, table));

      if (v.where) {
        query = query.where(buildWhere(toDrizzleColumn, v.where)) as any;
      }

      await query;
    },

    async create(table, values) {
      const idField = table.getIdColumn().names.drizzle;
      const drizzleTable = toDrizzle(table);
      values = mapValues(values, table);

      const returning: Record<string, ColumnType> = {};
      for (const column of Object.values(table.columns)) {
        returning[column.ormName] = drizzleTable[column.names.drizzle];
      }

      if (provider === "sqlite" || provider === "postgresql") {
        const result = await (db as unknown as P_DBType)
          .insert(drizzleTable as unknown as P_TableType)
          .values(values)
          .returning(returning as unknown as Record<string, P_ColumnType>);
        return result[0];
      }

      const obj = (
        await db.insert(drizzleTable).values(values).$returningId()
      )[0] as Record<string, unknown>;

      return (
        await db
          .select(returning)
          .from(drizzleTable)
          .where(Drizzle.eq(drizzleTable[idField], obj[idField]))
          .limit(1)
      )[0];
    },

    async createMany(table, values) {
      const idField = table.getIdColumn().names.drizzle;
      const drizzleTable = toDrizzle(table);
      values = values.map((v) => mapValues(v, table));

      if (provider === "sqlite" || provider === "postgresql") {
        return await (db as unknown as P_DBType)
          .insert(drizzleTable as unknown as P_TableType)
          .values(values)
          .returning({
            _id: (drizzleTable as unknown as P_TableType)[idField],
          });
      }

      const results: Record<string, unknown>[] = await db
        .insert(drizzleTable)
        .values(values)
        .$returningId();
      return results.map((result) => ({ _id: result[idField] }));
    },

    async deleteMany(table, v) {
      const drizzleTable = toDrizzle(table);
      let query = db.delete(drizzleTable);

      if (v.where) {
        query = query.where(buildWhere(toDrizzleColumn, v.where)) as any;
      }

      await query;
    },
    transaction(run) {
      return db.transaction((tx) => run(fromDrizzle(schema, tx, provider)));
    },
  });
}
