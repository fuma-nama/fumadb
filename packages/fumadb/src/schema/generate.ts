import { Schema } from "./create";

interface KyselyConfig {
  type: "kysely";
}

interface DrizzleConfig {
  type: "drizzle-orm";
}

interface TypeORMConfig {
  type: "type-orm";
}

export type Config = KyselyConfig | DrizzleConfig | TypeORMConfig;

/**
 * Generate schema for different ORMs.
 *
 * We don't want to implement the migrator, it's best for us to leverage existing solutions.
 */
export async function generate(
  schema: Schema,
  config: Config
): Promise<string> {
  return "";
}
