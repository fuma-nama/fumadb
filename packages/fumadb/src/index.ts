import { abstractQuery } from "./query";
import { Schema } from "./schema/create";

export function fumadb<S extends Schema>(schema: S) {
  return {
    get abstract() {
      return abstractQuery(schema);
    },
  };
}
