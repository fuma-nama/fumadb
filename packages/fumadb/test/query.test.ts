import { afterEach, expect, test, vi } from "vitest";
import {
  kyselyTests,
  drizzleTests,
  prismaTests,
  resetDB,
  databases,
  resetMongoDB,
} from "./shared";
import { fumadb } from "../src";
import fs from "fs";
import path from "path";
import { generateSchema } from "../src/schema/generate";
import { AbstractQuery } from "../src/query";
import * as DrizzleKit from "drizzle-kit/api";
import { v1 } from "./query/schema-1";

const myDB = fumadb({
  namespace: "test",
  schemas: [v1],
});

vi.mock("../src/cuid", () => ({
  createId: vi.fn(() => "generated-cuid"),
}));

afterEach(() => {
  vi.restoreAllMocks();
});

async function testMongoDatabase(orm: AbstractQuery<typeof v1>) {
  const { messages, users } = orm.tables;

  expect(
    await orm.create(users, {
      id: "generated-cuid",
      name: "fuma",
    })
  ).toMatchInlineSnapshot(`
      {
        "id": "generated-cuid",
        "name": "fuma",
      }
    `);

  await orm.createMany(users, [
    {
      id: "alfon",
      name: "alfon",
    },
  ]);

  expect(
    await orm.findMany(users, {
      select: true,
      orderBy: [[users.name, "asc"]],
    })
  ).toMatchInlineSnapshot(`
      [
        {
          "id": "alfon",
          "name": "alfon",
        },
        {
          "id": "generated-cuid",
          "name": "fuma",
        },
      ]
    `);

  await orm.createMany(messages, [
    {
      user: "alfon",
      content: "Hello World 1 by alfon",
      id: "1",
    },
    {
      user: "alfon",
      content: "Hello World 2 by alfon",
      id: "2",
      mentionId: "1",
    },
    {
      user: "bob",
      content: "Sad by bob",
      id: "3",
    },
  ]);

  expect(
    await orm.findMany(users, {
      orderBy: [users.name, "asc"],
      join: (b) =>
        b.messages({
          join: (b) =>
            b.mentionedBy({
              join: (b) => b.author(),
            }),
        }),
    })
  ).toMatchInlineSnapshot(`
    [
      {
        "id": "alfon",
        "messages": [
          {
            "content": "Hello World 1 by alfon",
            "id": "1",
            "image": null,
            "mentionId": null,
            "mentionedBy": {
              "author": {
                "id": "alfon",
                "name": "alfon",
              },
              "content": "Hello World 2 by alfon",
              "id": "2",
              "image": null,
              "mentionId": "1",
              "parent": null,
              "user": "alfon",
            },
            "parent": null,
            "user": "alfon",
          },
          {
            "content": "Hello World 2 by alfon",
            "id": "2",
            "image": null,
            "mentionId": "1",
            "mentionedBy": null,
            "parent": null,
            "user": "alfon",
          },
        ],
        "name": "alfon",
      },
      {
        "id": "generated-cuid",
        "messages": [],
        "name": "fuma",
      },
    ]
  `);

  expect(await orm.count(users)).toMatchInlineSnapshot(`2`);

  const getBob = () =>
    orm.findFirst(users, { where: (b) => b(users.id, "=", "bob") });
  const upsertBob = (v: string) =>
    orm.upsert(users, {
      where: (b) => b(users.id, "=", "bob"),
      create: { id: "bob", name: v },
      update: { name: v },
    });

  await upsertBob("Bob is sad");
  expect(await getBob()).toMatchInlineSnapshot(`
    {
      "id": "bob",
      "name": "Bob is sad",
    }
  `);

  await upsertBob("Bob is happy");
  expect(await getBob()).toMatchInlineSnapshot(`
    {
      "id": "bob",
      "name": "Bob is happy",
    }
  `);

  expect(
    await orm.create(messages, {
      id: "image-test",
      user: "alfon",
      content: "test",
      image: new Uint8Array([1, 2, 3, 4, 5, 6, 7, 8]),
    })
  ).toMatchInlineSnapshot(`
    {
      "content": "test",
      "id": "image-test",
      "image": Uint8Array [
        1,
        2,
        3,
        4,
        5,
        6,
        7,
        8,
      ],
      "mentionId": null,
      "parent": null,
      "user": "alfon",
    }
  `);

  await orm.updateMany(messages, {
    where: (b) => b(messages.id, "=", "image-test"),
    set: {
      content: "updated-content",
    },
  });
}

async function testSqlDatabase(orm: AbstractQuery<typeof v1>) {
  const { messages, users } = orm.tables;
  expect(
    await orm.create(users, {
      id: "generated-cuid",
      name: "fuma",
    })
  ).toMatchInlineSnapshot(`
      {
        "id": "generated-cuid",
        "name": "fuma",
      }
    `);
  await orm.createMany(users, [
    {
      id: "alfon",
      name: "alfon",
    },
  ]);
  expect(
    await orm.findMany(users, {
      select: true,
      orderBy: [[users.name, "asc"]],
    })
  ).toMatchInlineSnapshot(`
      [
        {
          "id": "alfon",
          "name": "alfon",
        },
        {
          "id": "generated-cuid",
          "name": "fuma",
        },
      ]
    `);
  await orm.createMany(messages, [
    {
      user: "alfon",
      content: "Hello World 1 by alfon",
      id: "1",
    },
    {
      user: "alfon",
      content: "Hello World 2 by alfon",
      id: "2",
      mentionId: "1",
    },
  ]);
  expect(
    await orm.findMany(users, {
      orderBy: [users.name, "asc"],
      join: (b) =>
        b.messages({
          join: (b) =>
            b.mentionedBy({
              join: (b) => b.author(),
            }),
        }),
    })
  ).toMatchInlineSnapshot(`
    [
      {
        "id": "alfon",
        "messages": [
          {
            "content": "Hello World 1 by alfon",
            "id": "1",
            "image": null,
            "mentionId": null,
            "mentionedBy": {
              "author": {
                "id": "alfon",
                "name": "alfon",
              },
              "content": "Hello World 2 by alfon",
              "id": "2",
              "image": null,
              "mentionId": "1",
              "parent": null,
              "user": "alfon",
            },
            "parent": null,
            "user": "alfon",
          },
          {
            "content": "Hello World 2 by alfon",
            "id": "2",
            "image": null,
            "mentionId": "1",
            "mentionedBy": null,
            "parent": null,
            "user": "alfon",
          },
        ],
        "name": "alfon",
      },
      {
        "id": "generated-cuid",
        "messages": [],
        "name": "fuma",
      },
    ]
  `);
  expect(
    await orm.findMany(users, {
      orderBy: [users.name, "asc"],
      join: (b) =>
        b.messages({
          select: ["content"],
          limit: 1,
          where: (b) => b(messages.content, "contains", "alfon"),
          join: (b) => b.author(),
        }),
    })
  ).toMatchInlineSnapshot(`
      [
        {
          "id": "alfon",
          "messages": [
            {
              "author": {
                "id": "alfon",
                "name": "alfon",
              },
              "content": "Hello World 1 by alfon",
            },
          ],
          "name": "alfon",
        },
        {
          "id": "generated-cuid",
          "messages": [],
          "name": "fuma",
        },
      ]
    `);

  expect(await orm.count(users)).toMatchInlineSnapshot(`2`);

  const getBob = () =>
    orm.findFirst(users, { where: (b) => b(users.id, "=", "bob") });
  const upsertBob = (v: string) =>
    orm.upsert(users, {
      where: (b) => b(users.id, "=", "bob"),
      create: { id: "bob", name: v },
      update: { name: v },
    });

  await upsertBob("Bob is sad");
  expect(await getBob()).toMatchInlineSnapshot(`
    {
      "id": "bob",
      "name": "Bob is sad",
    }
  `);

  await upsertBob("Bob is happy");
  expect(await getBob()).toMatchInlineSnapshot(`
    {
      "id": "bob",
      "name": "Bob is happy",
    }
  `);

  expect(
    await orm.create(messages, {
      id: "image-test",
      user: "alfon",
      content: "test",
      image: new Uint8Array([1, 2, 3, 4, 5, 6, 7, 8]),
    })
  ).toMatchInlineSnapshot(`
    {
      "content": "test",
      "id": "image-test",
      "image": Uint8Array [
        1,
        2,
        3,
        4,
        5,
        6,
        7,
        8,
      ],
      "mentionId": null,
      "parent": null,
      "user": "alfon",
    }
  `);

  const results = await Promise.all([
    orm
      .transaction(async (tx) => {
        await tx.createMany(messages, [
          {
            id: "transaction-1",
            user: "alfon",
            content: "test message",
          },
          {
            id: "transaction-2",
            user: "bob",
            content: "haha",
          },
        ]);

        await tx.deleteMany(messages, {
          where: (b) => b(messages.user, "=", "alfon"),
        });

        expect(
          await tx.findMany(messages, {
            where: (b) => b(messages.user, "=", "bob"),
          })
        ).toMatchInlineSnapshot(`
          [
            {
              "content": "haha",
              "id": "transaction-2",
              "image": null,
              "mentionId": null,
              "parent": null,
              "user": "bob",
            },
          ]
        `);

        throw new Error("Rollback!");
      })
      .catch(() => null),
    orm.findMany(messages, { where: (b) => b(messages.user, "=", "alfon") }),
  ]);

  // transaction should not affect concurrent operations
  expect(results[1]).toMatchInlineSnapshot(`
    [
      {
        "content": "Hello World 1 by alfon",
        "id": "1",
        "image": null,
        "mentionId": null,
        "parent": null,
        "user": "alfon",
      },
      {
        "content": "Hello World 2 by alfon",
        "id": "2",
        "image": null,
        "mentionId": "1",
        "parent": null,
        "user": "alfon",
      },
      {
        "content": "test",
        "id": "image-test",
        "image": Uint8Array [
          1,
          2,
          3,
          4,
          5,
          6,
          7,
          8,
        ],
        "mentionId": null,
        "parent": null,
        "user": "alfon",
      },
    ]
  `);
}

for (const item of kyselyTests) {
  test(`query ksely (${item.provider})`, async () => {
    await resetDB(item.provider);
    const instance = myDB.configure({
      type: "kysely",
      db: item.db,
      provider: item.provider,
    });

    const migrator = await instance.createMigrator();
    await migrator.migrateToLatest().then((res) => res.execute());

    const orm = instance.abstract;
    await testSqlDatabase(orm);
  });
}

test("query mongodb", async () => {
  const mongodb = databases.find((db) => db.provider === "mongodb")!.create();
  await mongodb.connect();
  await resetMongoDB(mongodb);

  const instance = myDB.configure({
    type: "mongodb",
    client: mongodb,
  });

  const orm = instance.abstract;
  await testMongoDatabase(orm);

  await mongodb.close();
});

for (const item of drizzleTests) {
  test(`query drizzle (${item.provider})`, async () => {
    await resetDB(item.provider);

    const schemaPath = path.join(
      import.meta.dirname,
      `drizzle-schema.${item.provider}.ts`
    );
    const schemaCode = generateSchema(v1, {
      type: "drizzle-orm",
      provider: item.provider,
    });
    fs.writeFileSync(schemaPath, schemaCode);

    const schema = await import(schemaPath);
    const db = item.db(schema);

    if (item.provider === "postgresql") {
      const { apply } = await DrizzleKit.pushSchema(schema, db as any);
      await apply();
    } else if (item.provider === "mysql") {
      const { sql } = await import("drizzle-orm");
      const prev = await DrizzleKit.generateMySQLDrizzleJson({});
      const cur = await DrizzleKit.generateMySQLDrizzleJson(schema);
      const statements = await DrizzleKit.generateMySQLMigration(prev, cur);

      for (const statement of statements) {
        await (db as any).execute(sql.raw(statement));
      }
    } else {
      // they need libsql
      const { apply } = await DrizzleKit.pushSQLiteSchema(schema, db as any);
      await apply();
    }

    const instance = myDB.configure({
      type: "drizzle-orm",
      db,
      provider: item.provider,
    });

    await testSqlDatabase(instance.abstract);

    fs.rmSync(schemaPath);
  });
}

for (const item of prismaTests) {
  test(`query prisma (${item.provider})`, { timeout: Infinity }, async () => {
    const prismaClient = await item.init(v1);

    const instance = myDB.configure({
      type: "prisma",
      prisma: prismaClient,
      provider: item.provider,
      db:
        item.provider === "mongodb"
          ? databases.find((db) => db.provider === "mongodb")!.create()
          : undefined,
    });

    const orm = instance.abstract;

    if (item.provider === "mongodb") {
      await testMongoDatabase(orm);
    } else {
      await testSqlDatabase(orm);
    }

    await prismaClient.$disconnect();
  });
}
