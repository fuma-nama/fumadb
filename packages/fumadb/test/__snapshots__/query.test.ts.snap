// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`query 'mysql' 1`] = `
[
  {
    "id": "generated-cuid",
    "name": "fuma",
  },
]
`;

exports[`query 'postgresql' 1`] = `
[
  {
    "id": "generated-cuid",
    "name": "fuma",
  },
]
`;

exports[`query 'sqlite' 1`] = `
[
  {
    "id": "generated-cuid",
    "name": "fuma",
  },
]
`;
